dmutils/status.py

```diff
@@ -47,3 +48,66 @@ def enabled_since(date_string):
         return date_string
 
     return False
+
+
+class StatusError(Exception):
+    """A stub class to use when implementing additional checks for an app's _status endpoint. See API for example."""
+    message = ''
+
+    def __init__(self, message):
+        super().__init__(message)
+        self.message = message
+
+
+def get_app_status(data_api_client=None,
+                   search_api_client=None,
+                   ignore_dependencies=False,
+                   additional_checks=None):
+    """Generates output for `_status` endpoints on apps
+
+    :param current_app: The flask `current_app` global.
+    :param data_api_client: The app's data_api_client, if used.
+    :param search_api_client: The app's search_api_client, if used.
+    :param ignore_dependencies: Minimal endpoint checks only (i.e. the app is routable and disk space is fine).
+    :param additional_checks: A sequence of callables that return dicts of data to be injected into the final JSON
+                              response or raise StatusErrors if they need to log an error that should fail the
+                              check (this will cause the endpoint to return a 500).
+    :return: A dictionary describing the current state of the app with, at least, a 'status' key with a value of 'ok'
+             or 'error'.
+    """
+    error_messages = []
+    response_data = {'status': 'ok'}
+
+    disk_status = get_disk_space_status()
+    response_data['disk'] = f'{disk_status[0]} ({disk_status[1]}% free)'
+    if disk_status[0].lower() != 'ok':
+        error_messages.append(f'Disk space low: {disk_status[1]}% remaining.')
+
+    if not ignore_dependencies:
+        response_data['version'] = current_app.config['VERSION']
+        response_data['flags'] = get_flags(current_app)
+
+        if data_api_client:
+            data_api_status = data_api_client.get_status() or {'status': 'n/a'}
+            response_data['api_status'] = data_api_status
+            if data_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Data API.')
+
+        if search_api_client:
+            search_api_status = search_api_client.get_status() or {'status': 'n/a'}
+            response_data['search_api_status'] = search_api_status
+            if search_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Search API.')
+
+        for additional_check in (additional_checks or []):
+            try:
+                response_data.update(additional_check())
+
+            except StatusError as e:
+                error_messages.append(e.message)
```

Comment:
(This is all great btw.) So the question remaining in my head is whether we should be making sure an indication of _what_ additional check was being performed, enforced as being in the json output, somehow.

And a follow-on question is about what happens if additional_check causes some other error to occur, but I suppose as long as the end result in that case is a 500, then it's probably fine to not be catching those.

Reply:
Yeah, at the moment all of the additional checks just output 500s. If we need extra granularity in the future we can update this to take a message and a status code, but for now it's not needed, so I left it out.

Reply:
Updated the docstring on `StatusError`.